packages/compose/src/hooks/use-shallow-compare-effect/index.js

```diff
@@ -0,0 +1,26 @@
+/**
+ * WordPress dependencies
+ */
+import { useEffect, useRef } from '@wordpress/element';
+import isShallowEqual from '@wordpress/is-shallow-equal';
+
+/**
+ * Like `useEffect` but call the effect when the dependencies are not shallowly equal.
+ * Useful when the size of the dependency array might change during re-renders.
+ * This hook is only used for backward-compatibility reason. Consider using `useEffect` wherever possible.
+ *
+ * @param {Function} effect The effect callback passed to `useEffect`.
+ * @param {Array} deps The dependency array that is compared against shallowly.
+ */
+function useShallowCompareEffect( effect, deps ) {
```

Comment:
This hook is a bit confusing what's the difference between doing this and spreading `deps` in the dependencies?

Reply:
The main difference is that in the original effect, we are passing a dependency array dynamically, which means the size of the array could change. When it changes, it will throw a warning from React like below.

> Warning: The final argument passed to useEffect changed size between renders. The order and size of this array must remain constant.

This hook _hack_ around this by putting them inside an item in the dependency array, so that the even when the size changed, it won't throw warnings at us.